mmcv/runner/ipu/util.py

```diff
@@ -0,0 +1,265 @@
+
+    # set model partition
+    if optimizer is None:
+        train_model = None
+    else:
+        # split model into multi-IPUs if specified
+        train_model = model_sharding(
+            copy.copy(model).train(),
+            ipu_model_cfg.get('train_split_edges', []))
+
+        recomputation_checkpoint(
+            train_model,
+            ipu_model_cfg.get('train_ckpt_nodes', []))
+
+        # TODO support feature alignment for gradient accumulation mode
+        if getattr(opts['training'].Training, 'gradient_accumulation', 1) > 1:
+            assert len(modules_to_record) == 0, \
+                'Feature alignment for grad-accumulation mode not implemented'
+
+        # TODO support feature alignment for multi-replica mode
+        if getattr(opts['training'], 'replication_factor', 1) > 1:
+            assert len(modules_to_record) == 0, \
+                'Feature alignment for multi-replica mode not implemented'
+
+    # split model into multi-IPUs if specified
+    assert len(ipu_model_cfg.get('eval_split_edges', [])) == 0,\
+        'Currently, BeginBlock can only be used once on the same model'
+    eval_model = model_sharding(copy.copy(model).eval(), ipu_model_cfg.get(
+        'eval_split_edges', []))
+
+    # wrap model for compilation
+    model = TrainEvalModel(train_model, eval_model, options=opts,
+                           optimizer=optimizer, logger=logger,
+                           modules_to_record=modules_to_record)
+    model.train(training)
+    return model
+
+
+def model_sharding(model, split_edges):
+    """split models in-place into multi-IPUs
+
+    Args:
+        model (pytorch.nn.Module): the target model to be split
+        split_edges (dict): model layer names or layer numbers
+            of split edge
+
+    Returns:
+        model (pytorch.nn.Module): split model
+    """
+    if len(split_edges) == 0:
+        return model
+    assert isinstance(split_edges, list)
+    spilt_edges_dic = {ele['layer_to_call']: ele for ele in split_edges}
+    for idx, (_name, _module) in enumerate(model.named_modules()):
+        assert not (idx in spilt_edges_dic and _name in spilt_edges_dic),\
+            f'The same layer is referenced twice while doing model partition: \
+                idx is {idx} and name is {_name}'
+        edge = spilt_edges_dic.pop(_name, None)
+        edge = spilt_edges_dic.pop(idx, edge)
+        if edge is not None:
+            poptorch.BeginBlock(_module, edge.get(
+                'user_id', _name), edge['ipu_id'])
+    # check all split_edges are used
+    split_edge_names = list(spilt_edges_dic.keys())
+    assert len(spilt_edges_dic) == 0,\
+        f'split_edges: {split_edge_names} are not contained in the model'
+    return model
+
+
+def recomputation_checkpoint(model: torch.nn.Module, module_names=[])\
+     -> torch.utils.hooks.RemovableHandle:
+    """Annotates the output of a module to be checkpointed instead of
+        recomputed"""
```

Comment:
```suggestion
    """Annotates the output of a module to be checkpointed instead of
        recomputed.
        
    Args:
        model (nn.Module): ...
        module_names (list): ...
    """
```